File: windows/Hingewave.Core/LaptopMotionModel.cs
namespace Hingewave.Core;

public enum MotionState
{
    Idle,
    Armed,
    Active,
    Clearing,
}

/// <summary>One frame of the motion model.</summary>
/// <param name="State">Current state.</param>
/// <param name="Tilt">Degrees the panel has rotated away from the resting plane.</param>
/// <param name="Progress">0 open, 1 fully folded.</param>
/// <param name="Capture">The host should be capturing the screen.</param>
public readonly record struct MotionOutput(MotionState State, double Tilt, double Progress, bool Capture)
{
    public static readonly MotionOutput Idle = new(MotionState.Idle, 0.0, 0.0, false);
}

public static class Easing
{
    public static double Smoothstep(double x)
    {
        if (x <= 0.0) return 0.0;
        if (x >= 1.0) return 1.0;
        return x * x * (3.0 - 2.0 * x);
    }
}

/// <summary>
/// Motion model for laptop lids. Executable form of docs/design.md section 3.2,
/// mirroring core/reference/hingewave_ref/motion.py and verified by core/fixtures.
/// </summary>
public sealed class LaptopMotionModel
{
    private const double StillSpeed = 2.0;
    private const double EndMargin = 2.0;
    /// <summary>Within this many degrees of the rest angle counts as reopened.</summary>
    private const double ReopenMargin = 2.0;

    public EffectConfig Config { get; }
    public bool ReduceMotion { get; set; }
    public MotionState State { get; private set; } = MotionState.Idle;
    public bool CaptureSeen { get; private set; }

    /// <summary>The smoothed angle after the last feed, for display.</summary>
    public double SmoothedAngle => _spring.X;

    private readonly Spring _spring;
    private double? _lastT;
    private double? _stillSince;
    private double? _clearStarted;
    private double? _restStillSince;

    /// <summary>Angle the lid rests at; follows the lid whenever it has been still for RestSeconds while idle.</summary>
    public double? RestAngle { get; private set; }

    public LaptopMotionModel(EffectConfig? config = null, bool reduceMotion = false)
    {
        Config = config ?? EffectConfig.Defaults;
        ReduceMotion = reduceMotion;
        _spring = new Spring(Config.SpringHz);
    }

    public void CaptureReady(double t)
    {
        CaptureSeen = true;
        if (State == MotionState.Armed) State = MotionState.Active;
    }

    public void DisplayOff() => GoIdle();

    public MotionOutput Feed(double t, double angle)
    {
        double x, v;
        if (_lastT is double last)
        {
            (x, v) = _spring.Step(angle, t - last);
        }
        else
        {
            _spring.Reset(angle);
            x = angle;
            v = 0.0;
        }
        _lastT = t;

        var lp = Config.Laptop;

        if (State == MotionState.Idle)
        {
            RestAngle ??= x;
            if (Math.Abs(v) < StillSpeed)
            {
                if (_restStillSince is double since)
                {
                    if (t - since >= lp.RestSeconds) RestAngle = x;
                }
                else
                {
                    _restStillSince = t;
                }
            }
            else
            {
                _restStillSince = null;
            }
            if (RestAngle is double rest && rest - x >= lp.ArmDelta && v <= -lp.ArmVelocity)
            {
                State = MotionState.Armed;
                CaptureSeen = false;
                _stillSince = null;
            }
        }

        if (State is MotionState.Armed or MotionState.Active)
        {
            if (RestAngle is double r && x > r - ReopenMargin)
            {
                BeginClearing(t);
            }
            else if (Math.Abs(v) < StillSpeed && x > lp.EndAngle + EndMargin)
            {
                if (_stillSince is double since)
                {
                    if (t - since >= Config.StillSeconds) BeginClearing(t);
                }
                else
                {
                    _stillSince = t;
                }
            }
            else
            {
                _stillSince = null;
            }
        }

        if (State == MotionState.Idle) return MotionOutput.Idle;

        var (tilt, progress) = Shape(x);

        if (State == MotionState.Clearing && _clearStarted is double started)
        {
            var k = 1.0 - (t - started) / Config.ClearSeconds;
            if (k <= 0.0)
            {
                GoIdle();
                return MotionOutput.Idle;
            }
            tilt *= k;
            progress *= k;
        }

        return new MotionOutput(State, tilt, progress, true);
    }

    private (double Tilt, double Progress) Shape(double x)
    {
        var lp = Config.Laptop;
        var rest = RestAngle ?? x;
        var tilt = ReduceMotion ? 0.0 : Math.Max(0.0, rest - x);
        var progress = Easing.Smoothstep((rest - x) / Math.Max(rest - lp.EndAngle, 1e-6));
        return (tilt, progress);
    }

    private void BeginClearing(double t)
    {
        if (State == MotionState.Clearing) return;
        State = MotionState.Clearing;
        _clearStarted = t;
        _stillSince = null;
    }

    private void GoIdle()
    {
        State = MotionState.Idle;
        _stillSince = null;
        _clearStarted = null;
        CaptureSeen = false;
        // Re-anchor at the current angle so a fresh close starts from here.
        RestAngle = null;
        _restStillSince = null;
    }
}

File: macos/Sources/HingewaveCore/LaptopMotionModel.swift
import Foundation

/// Motion model for laptop lids. Executable form of docs/design.md section 3.2,
/// mirroring core/reference/hingewave_ref/motion.py and verified by core/fixtures.
public enum MotionState: String, Codable {
    case idle, armed, active, clearing
}

public struct MotionOutput: Equatable {
    public var state: MotionState
    /// Degrees the panel has rotated away from the resting plane.
    public var tilt: Double
    /// 0 open, 1 fully folded.
    public var progress: Double
    /// The host should be capturing the screen.
    public var capture: Bool

    public static let idle = MotionOutput(state: .idle, tilt: 0, progress: 0, capture: false)
}

public func smoothstep(_ x: Double) -> Double {
    if x <= 0 { return 0 }
    if x >= 1 { return 1 }
    return x * x * (3 - 2 * x)
}

public struct LaptopMotionModel {
    static let stillSpeed = 2.0
    static let endMargin = 2.0

    public let config: EffectConfig
    public var reduceMotion: Bool
    public private(set) var state: MotionState = .idle
    public private(set) var captureSeen = false

    private var spring: Spring
    private var lastT: Double?
    private var prevX: Double?
    private var stillSince: Double?
    private var clearStarted: Double?

    public init(config: EffectConfig = .defaults, reduceMotion: Bool = false) {
        self.config = config
        self.reduceMotion = reduceMotion
        self.spring = Spring(hz: config.springHz)
    }

    /// The smoothed angle after the last feed, for display.
    public var smoothedAngle: Double { spring.x }

    // MARK: Host events

    public mutating func captureReady(t: Double) {
        captureSeen = true
        if state == .armed { state = .active }
    }

    public mutating func displayOff() {
        goIdle()
    }

    // MARK: Sampling

    public mutating func feed(t: Double, angle: Double) -> MotionOutput {
        let x: Double
        let v: Double
        if let last = lastT {
            (x, v) = spring.step(target: angle, dt: t - last)
        } else {
            spring.reset(angle)
            x = angle
            v = 0
        }
        lastT = t
        let prev = prevX
        prevX = x

        let lp = config.laptop

        if state == .idle {
            let crossed = prev.map { $0 >= lp.startAngle && x < lp.startAngle } ?? false
            if crossed && v <= -lp.armVelocity {
                state = .armed
                captureSeen = false
                stillSince = nil
            }
        }

        if state == .armed || state == .active {
            if x > lp.startAngle {
                beginClearing(t: t)
            } else if abs(v) < Self.stillSpeed && x > lp.endAngle + Self.endMargin {
                if let since = stillSince {
                    if t - since >= config.stillSeconds { beginClearing(t: t) }
                } else {
                    stillSince = t
                }
            } else {
                stillSince = nil
            }
        }

        if state == .idle { return .idle }

        var (tilt, progress) = shape(x)

        if state == .clearing, let started = clearStarted {
            let k = 1.0 - (t - started) / config.clearSeconds
            if k <= 0 {
                goIdle()
                return .idle
            }
            tilt *= k
            progress *= k
        }

        return MotionOutput(state: state, tilt: tilt, progress: progress, capture: true)
    }

    // MARK: Internals

    private func shape(_ x: Double) -> (Double, Double) {
        let lp = config.laptop
        let tilt = reduceMotion ? 0 : max(0, lp.startAngle - x)
        let progress = smoothstep((lp.startAngle - x) / (lp.startAngle - lp.endAngle))
        return (tilt, progress)
    }

    private mutating func beginClearing(t: Double) {
        guard state != .clearing else { return }
        state = .clearing
        clearStarted = t
        stillSince = nil
    }

    private mutating func goIdle() {
        state = .idle
        stillSince = nil
        clearStarted = nil
        captureSeen = false
    }
}
